gflownet/envs/crystals/crystal.py

```diff
@@ -0,0 +1,488 @@
+        # Triclinic was used because it doesn't force any initial starting angles.
+        self.lattice_parameters = LatticeParameters(
+            lattice_system=TRICLINIC, **self.lattice_parameters_kwargs
+        )
+
+        # 0-th element of state encodes current stage: 0 for composition,
+        # 1 for space group, 2 for lattice parameters
+        self.source = (
+            [0]
+            + self.composition.source
+            + self.space_group.source
+            + self.lattice_parameters.source
+        )
+
+        # start and end indices of individual substates
+        self.composition_state_start = 1
+        self.composition_state_end = self.composition_state_start + len(
+            self.composition.source
+        )
+        self.space_group_state_start = self.composition_state_end
+        self.space_group_state_end = self.space_group_state_start + len(
+            self.space_group.source
+        )
+        self.lattice_parameters_state_start = self.space_group_state_end
+        self.lattice_parameters_state_end = self.lattice_parameters_state_start + len(
+            self.lattice_parameters.source
+        )
+
+        # start and end indices of individual submasks
+        self.composition_mask_start = 0
+        self.composition_mask_end = self.composition_mask_start + len(
+            self.composition.action_space
+        )
+        self.space_group_mask_start = self.composition_mask_end
+        self.space_group_mask_end = self.space_group_mask_start + len(
+            self.space_group.action_space
+        )
+        self.lattice_parameters_mask_start = self.space_group_mask_end
+        self.lattice_parameters_mask_end = self.lattice_parameters_mask_start + len(
+            self.lattice_parameters.action_space
+        )
+
+        self.eos = self.lattice_parameters.eos
+        self.stage = Stage.COMPOSITION
+        self.max_action_length = max(
+            max(len(a) for a in self.composition.action_space),
+            max(len(a) for a in self.space_group.action_space),
+            max(len(a) for a in self.lattice_parameters.action_space),
+        )
+        self.done = False
+
+        # Conversions
+        self.state2proxy = self.state2oracle
+        self.statebatch2proxy = self.statebatch2oracle
+        self.statetorch2proxy = self.statetorch2oracle
+
+        super().__init__(**kwargs)
+
+    def _set_lattice_parameters(self):
+        """
+        Sets LatticeParameters conditioned on the lattice system derived from the SpaceGroup.
+        """
+        crystal_system = self.space_group.get_crystal_system()
+
+        if crystal_system is None:
+            raise ValueError(
+                "Cannot set lattice parameters without crystal system determined in the space group."
+            )
+
+        self.lattice_parameters = LatticeParameters(
+            lattice_system=self.space_group.lattice_system,
+            **self.lattice_parameters_kwargs,
+        )
```

Comment:
Shouldn't the lattice parameters work with crystal-lattice system instead of crystal system?

Reply:
I am not sure what you mean? `space_group.lattice_system` is being used currently in the code you cited, so that's not the crystal system, right?

Reply:
(I think the confusion might have been, I was checking for `crystal_system` being set and raised error if it wasn't, but the `LatticeParameters` were, I think properly, conditioned on the `lattice_system`)